### ISSUE TITLE
Serialization exception when using the internal prediction API endpoint

### ISSUE THREAD

**Issue by pisymbol** [SeldonIO/seldon-core#1241]
In an effort to work around the lack of NGINX support, I am trying to redirect model requests to use Seldon Core's internal API (/api/v0.1/predictions endpoint). However, when I do that, it breaks on inference return (the end of my predict() function using the Python wrapper API) with:

```
com.google.protobuf.InvalidProtocolBufferException: Failed parsing JSON source: JsonReader at line 1 column 70566875 path $.data.tensor.values[20767726] to Json
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1347) ~[protobuf-java-util-3.10.0.jar!/:?]
	at com.google.protobuf.util.JsonFormat$Parser.merge(JsonFormat.java:477) ~[protobuf-java-util-3.10.0.jar!/:?]
	at io.seldon.engine.service.InternalPredictionService.queryREST(InternalPredictionService.java:498) ~[classes!/:0.5.1]
	at io.seldon.engine.service.InternalPredictionService.transformInput(InternalPredictionService.java:291) ~[classes!/:0.5.1]
```

This is Seldon Core 0.5.1.

As a test, I wrote a model that simply converts the input image at predict time (basically converts it to a numpy array and returns it as the result).

Some sample code:

```
from PIL import Image
import numpy as np

# X here is the input tensor passed into the predict() call
img = Image.open(io.BytesIO(X)).convert('RGB')
img = np.array(img)

# EDIT: Convert to BGR
img = img[:,:,::-1]

# EDIT: Return RGB.
img = img[:,:,::-1]

return img
```

What am I doing wrong?

**Comment by cliveseldon** [comment]
This looks like the JSON has been corrupted maybe due to the size of the response. Can you run the engine in DEBUG mode to check the JSON returned? See [here](https://docs.seldon.io/projects/seldon-core/en/latest/analytics/log_level.html#setting-log-level-in-the-seldon-engine).

The this line should show the response: https://github.com/SeldonIO/seldon-core/blob/5bf3894fdedcc5c7e67f792d9408a26b2ad5c0e2/engine/src/main/java/io/seldon/engine/service/InternalPredictionService.java#L496

**Comment by pisymbol** [comment]
Looks OK to me:

```
2019-12-11 15:07:27.227 DEBUG 6 --- [         task-2] i.s.e.s.InternalPredictionService        : {"data":{"names":["t:0",
...
,"tensor":{"shape":[4032,3024,3],"values":[255,255,255...
,71,72,99,83,58,85,69,53,80,64,52,79,63,60,87,71,60,87,71,62,89,73,58,87,71,59,88,72,61,90,74,61,90,74,59,86,70,58,85,69,53,80,64,42,69,53,49,74,60,47,72,58,58,83,69,52,75,61,56,79,65,49,72,58,55,78,64,45,68,54,51,76,62,49,74,60,66,91,77,46,71,57,39,64,50,46,71,57,50,75,61,48,73,59,46,71,57,46,71,57,47,72,58,45,70,56,45,70,56,42,67,53,37,62,48,44,69,55,41,66,52,40,65,51,40,65,51,41,66,52,41,67,51,40,66,50,38,64,48,36,62,46,39,62,47,41,64,49,39,62,47,34,57,42,35,58,43,39,62,47,39,62,47,36,59,44,38,61,47,36,59,45,35,58,44,35,58,44,39,62,48,33,56,42,25,48,34,36,59,45,32,55,41,36,59,45,34,57,43,36,59,45,34,57,43,30,53,39,34,57,43,32,55,41,34,56,44,34,56,44,33,55,43,33,55,43,33,55,43,33,55,43,34,56,44,34,56,44,33,56,42,33,56,42,36,59,45,33,56,42,35,58,44,39,62,48,35,58,44,32,55,41,32,54,42,33,55,43,33,55,43,33,55,43,32,54,42,32,55,41,33,56,42,34,57,43,34,57,42,28,51,36,36,60,42,34,58,40,18,42,24,40,64,46,53,78,58,44,69,49,39,67,44,24,52,29,43,72,49,45,74,51,48,78,53,50,82,57,54,89,63,55,90,64,53,91,63,49,87,59,53,93,65,62,103,75,63,105,74,64,106,75,64,108,77,60,104,73,62,110,81,59,108,76,69,116,84,80,124,93,78,121,88,78,120,85,82,124,89,82,123,86,84,122,86,85,124,86,86,125,87,89,128,90,90,131,93,92,133,95,94,135,97,95,136,98,92,133,95,92,133,95,93,134,96,91,135,96,89,133,92,87,131,90,89,133,92,92,136,95,96,142,100,99,144,105,101,146,107,101,146,107,98,145,107,97,144,106,98,145,107,99,146,108,96,143,104,97,144,105,97,144,105,97,144,105,97,144,105,100,147,108,104,151,112,108,156,114,111,159,117,109,157,115,108,156,114,108,156,114,109,157,115,109,157,115,109,157,115,108,156,114,108,157,113,109,158,114,109,158,114,109,158,114,109,158,114,109,157,115,111,159,117,112,160,118,109,159,119,109,159,119,110,160,120,111,161,121,112,161,123,114,163,125,116,165,127,116,165,127]}},"meta":{"metrics":[{"key":"total_predicts","type":"COUNTER","value":1},{"key":"total_predict_time","type":"COUNTER","value":0.2662653923034668},{"key":"predict_time_avg","type":"TIMER","value":266.2653923034668}],"puid":"piokrp3o5so8ef7himobn704ar"}}

2019-12-11 15:07:30.406 DEBUG 6 --- [   scheduling-1] i.s.e.a.r.SeldonGraphReadyChecker        : Seldon graph ready: true
2019-12-11 15:07:35.407 DEBUG 6 --- [   scheduling-1] i.s.e.a.r.SeldonGraphReadyChecker        : Seldon graph ready: true
2019-12-11 15:07:40.408 DEBUG 6 --- [   scheduling-1] i.s.e.a.r.SeldonGraphReadyChecker        : Seldon graph ready: true
2019-12-11 15:07:48.114 DEBUG 6 --- [   scheduling-1] i.s.e.a.r.SeldonGraphReadyChecker        : Seldon graph ready: true
2019-12-11 15:07:48.114 DEBUG 6 --- [         task-2] i.s.e.s.InternalPredictionService        : External prediction server took 32988ms
2019-12-11 15:07:48.114 ERROR 6 --- [         task-2] i.s.e.s.InternalPredictionService        : Invalid protocol buffer during Json Format merge -

com.google.protobuf.InvalidProtocolBufferException: Failed parsing JSON source: JsonReader at line 1 column 70566875 path $.data.tensor.values[20767726] to Json
	at com.google.protobuf.util.JsonFormat$ParserImpl.merge(JsonFormat.java:1347) ~[protobuf-java-util-3.10.0.jar!/:?]
	at com.google.protobuf.util.JsonFormat$Parser.merge(JsonFormat.java:477) ~[protobuf-java-util-3.10.0.jar!/:?]
	at io.seldon.engine.service.InternalPredictionService.queryREST(InternalPredictionService.java:498) ~[classes!/:0.5.1]
	at io.seldon.engine.service.InternalPredictionService.transformInput(InternalPredictionService.java:291) ~[classes!/:0.5.1]
	at io.seldon.engine.predictors.PredictiveUnitBean.transformInput(PredictiveUnitBean.java:257) ~[classes!/:0.5.1]
	at io.seldon.engine.predictors.PredictiveUnitBean.getOutputAsync(PredictiveUnitBean.java:134) ~[classes!/:0.5.1]
	at io.seldon.engine.predictors.PredictiveUnitBean$$FastClassBySpringCGLIB$$ddb0d27b.invoke(<generated>) ~[classes!/:0.5.1]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:769) ~[spring-aop-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:747) ~[spring-aop-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.2.0.RELEASE.jar!/:5.2.0.RELEASE]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_222]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_222]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
```

**Comment by cliveseldon** [comment]
Are you able to create a Unit Test that uses that content if you can save in file?

**Comment by pisymbol** [comment]
You mean if the return is parseable by say the standard Python json module? Is that what you are asking?

**Comment by cliveseldon** [comment]
No. Save the JSON in a file and create a unit test in the engine that shows the error or not.

**Comment by cliveseldon** [comment]
Calling the protobuf parse in the unit test.

**Comment by pisymbol** [comment]
I can try. But what changed? Note that this worked fine in 0.3.2-SNAPSHOT (same exact code and image (the image's pixel values changed but the size and structure of the numpy array was identical)).

**Comment by pisymbol** [comment]
FYI: If I send a smaller image, it works. Definitely seems like some kind of size limit when parsing.

**Comment by adriangonz** [comment]
This may be a red herring, but since it's complaining about the `$.data.tensor.values[20767726]` JSON path, is it relevant that the final positions of the `values` array seem to be "empty" commas?

```
... "values":[255,255,255...,71,7246,,,] ...
```

**Comment by pisymbol** [comment]
I didn't check the types of both images (pretty sure they are RGB, i.e. a three channel tensor).

Well, the JSON handed off to the protobuf parser is constructed in python/seldon_core/utils.py in construct_response_json(), right?

It looks like the image tensor is simply returned as:

img.ravel().tolist()

which should return a 1-D Python list which I would assume is JSON serializeable, right?

The bottom line @adriangonz is that AFAICT, I don't touch the values of the image This is all in your encoding scheme.

**Comment by pisymbol** [comment]
Well, I take back my size comment. It is true a smaller jpeg worked (~76k) but I just sent an over 3.4M png at my service and that returned HTTP OK and worked fine.

**Comment by pisymbol** [comment]
If I write out the DEBUG output to a file as per @cliveseldon and load it back using at least Python's built in JSON parser it works fine. 

But here's something strange:

In [4]: len(js['data']['tensor']['values'])
Out[4]: 7406630

So 'js' here is the raw JSON displayed by the logger.debug() message after the seldon-container-engine collects the raw response from the Python wrapper (i.e. the return from my predict function).

But if I read the image and np.ravel().tolist() it I see:

n [16]: img.shape
Out[16]: (4032, 3024, 3)

In [17]: len(np.ravel(img).tolist())
Out[17]: 36578304

Which makes sense since it is 4032 x 3024 x 3. 

It looks like values got truncated somehow????

**Comment by cliveseldon** [comment]
I wonder if there is a max buffer size for JSON REST at the send or receive side?

**Comment by pisymbol** [comment]
setMaxPostSize()?

https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#common-application-properties

Just looking at maybe Spring is limiting the size when sending to predict?

**Comment by pisymbol** [comment]
So @cliveseldon I focused on the python wrapper since that is the sender back to the container engine - the jsonify call as part of construct_response_json() looks good. I wrote out the response json as a file in the container and then inspected it (logger.debug() here is too awkward given the size of the JPEG):

In [8]: len(js['data']['tensor']['values'])
Out[8]: 36578304
In [9]: 4032*3024*3
Out[9]: 36578304

So the response back is well formed JSON. But at the receiver's side (the Java side) within queryREST() the httpBody seems truncated (I will have to double check though).

**Comment by pisymbol** [comment]
EUREKA! Java is running out of heap space!

**Comment by pisymbol** [comment]
The fix is to raise heap under Java so it can serialize the JSON to a protobuf. Can we please, PLEASE raise the default heap space allocation for 0.6? It is very unfriendly right now if you are sending and receiving large images because of weird errors like this one.

What's odd is the Java process didn't restart after running out of heap so this one was harder to track than usual (the last time I had a mem usage error the whole container crashed which made it more obvious).

**Comment by cliveseldon** [comment]
You can set the JAVA_OPTS to set the heap space for the engine from the defaults: https://docs.seldon.io/projects/seldon-core/en/latest/graph/svcorch.html#java-settings

**Comment by pisymbol** [comment]
But do you think a user should have to do that? Like ever.... :-)

**Comment by pisymbol** [comment]
The other thing I want ot mention is the fix that worked from one service did not work for the other. So now I'm trying to understand how to profile the container engine to understand how much heap it needs per service. This is not very user friendly unless you think I'm doing something wrong.

**Comment by cliveseldon** [comment]
I don't think we can know how much heap is needed for your ML Deployment. So we could increase the default if that is too low but then you are penalizing users with slim deployments. I think realisitically you do need to profile your application. 

This may be less of an issue with the Go executor replacement for the Java engine.

**Comment by pisymbol** [comment]
Arrgh, another model is throwing this now: {'code': 203, 'info': 'com.google.protobuf.InvalidProtocolBufferException: Failed parsing JSON source: JsonReader at line 1 column 101319540 path $.data.tensor.values[18682580] to Json', 'reason': 'Microservice error', 'status': 'FAILURE'}

Technical note:

The problem with profiling the seldon-container-engine is that it requires Java expertise as well as a bunch of tools that are probably not in the purview of most MLE/DSs.

In terms of heap space penalizing slim deployments, maybe just setting the max heap size to something higher is all that is needed (that way the JVM's GC can scale accordingly).

I am experimenting with the -Xmx setting.

**Comment by pisymbol** [comment]
The other question I have is why is this not an issue with /predict on 9000?
